src/Simulation/Common/QubitManager.cs

```diff
@@ -74,185 +171,224 @@ public class QubitManager
-        private bool DecreaseBorrowingRefCount(long id)
+        public long GetFreeQubitsCount()
         {
-            if (IsAllocatedForBorrowing(id)) {
-                if (qubits[id] == AllocatedForBorrowing)
-                {
-                    return true;
-                }
-                else
-                {
-                    qubits[id]--;
-                    return false;
-                }
-            } else
+            return NumQubits - NumDisabledQubits - NumAllocatedQubits;
+        }
+
+        public long GetAllocatedQubitsCount()
+        {
+            return NumAllocatedQubits;
+        }
+
+        public virtual IEnumerable<long> GetAllocatedIds()
+        {
+            for (long i = 0; i < qubits.LongLength; i++)
             {
-                return false;
+                if (!IsFree(i)) yield return i;
             }
         }
 
-        private void IncreaseBorrowingRefCount(long id)
+        #region Disable
+
+        /// <summary>
+        /// Disables a given qubit.
+        /// Once a qubit is disabled it can never be reallocated.
+        /// </summary>
+        public virtual void Disable(Qubit qubit)
+        {
+            qubits[qubit.Id] = Disabled;
+            NumDisabledQubits++;
+
+            NumAllocatedQubits--;
+            Debug.Assert(NumAllocatedQubits >= 0);
+        }
+
+        /// <summary>
+        /// Disables a set of given qubits.
+        /// Once a qubit is disabled it can never be reallocated.
+        /// </summary>
+        public void Disable(IQArray<Qubit> qubitsToDisable)
         {
-            if (IsAllocatedForBorrowing(id))
+            if (qubitsToDisable == null || qubitsToDisable.Length == 0)
             {
-                qubits[id]++;
+                return;
+            }
+
+            foreach (Qubit qubit in qubitsToDisable)
+            {
+                this.Disable(qubit);
             }
         }
 
+        #endregion
+
+        #region Allocate
+
         /// <summary>
         /// Allocates a qubit.
+        /// Returns null if the qubit cannot be allocated.
         /// </summary>
-        protected virtual Qubit AllocateOneQubit(bool usedOnlyForBorrowing)
+        protected virtual Qubit Allocate(bool usedOnlyForBorrowing)
         {
             if (free == None)
             {
-                if (MayExtendCapacity)
+                if (!MayExtendCapacity)
                 {
-                    ExtendQubitArray();
+                    return null;
                 }
-                else
+
+                long oldNumQubits = NumQubits;
+                long oldNone = None;
+                long oldAllocated = Allocated;
+                long oldDisabled = Disabled;
+                long oldAllocatedForBorrowing = AllocatedForBorrowing;
+                long[] oldQubitsArray = this.qubits;
+
+                UpdateQubitCapacity(oldNumQubits * 2); // This changes Allocated, Disabled, AllocatedForBorrowing markers.
```

Comment:
When allocating a single new qubit, why do we update the capacity by 2x?

Reply:
I believe the capacity is only doubled if if free == None (meaning no more qubits can be allocated without exceeding the current capacity), and MayExtendCapacity is true.

Reply:
Ah, I see that now! Thanks!